Feature/puneet/migrateandroid x bookmarks

puneet-pdx opened the pull request.

LDuncAndroid commented on arcgis-android-toolkit/src/main/java/com/esri/arcgisruntime/toolkit/bookmark/BookmarkAdapter.kt:
```diff
@@ -0,0 +1,76 @@
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import androidx.annotation.LayoutRes
```
This is no longer used. Is there any reason you removed the `@LayoutRes` annotation?

puneet-pdx replied:
no reason, I just forgot to apply it when moving things around. I have added it now.

LDuncAndroid commented on arcgis-android-toolkit/src/main/java/com/esri/arcgisruntime/toolkit/bookmark/BookmarkAdapter.kt:
```diff
@@ -0,0 +1,76 @@
+class BookmarkAdapter(
+    private val onItemClickListener: OnItemClickListener<Bookmark>,
+    diffCallback: DiffUtil.ItemCallback<Bookmark> = DiffCallback(),
+    private val itemLayoutRes: Int? = null
```
You can get rid of the null here and the null check further down by providing a default optional parameter in this constructor.

```
class BookmarkAdapter(
    private val onItemClickListener: OnItemClickListener<Bookmark>,
    @LayoutRes private val  itemLayoutRes: Int = R.layout.item_bookmark_row,
    diffCallback: DiffUtil.ItemCallback<Bookmark> = DiffCallback()
) : ListAdapter<Bookmark, ViewHolder>(diffCallback) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        return ViewHolder(
            inflater.inflate(
                itemLayoutRes,
                parent,
                false
            )
        )
    }
```

And calling the constructor would look like:
```
bookmarksAdapter = BookmarkAdapter(
                object : BookmarkAdapter.OnItemClickListener<Bookmark> {
                    override fun onItemClick(item: Bookmark) {
                        onItemClickListener?.onItemClick(item)
                    }
                },
                itemLayoutRes = R.layout.your_own_layout
            )
```

puneet-pdx replied:
updated

LDuncAndroid commented on arcgis-android-toolkit/src/main/java/com/esri/arcgisruntime/toolkit/bookmark/BookmarkAdapter.kt:
```diff
@@ -0,0 +1,76 @@
+        bookmark: Bookmark,
+        onItemClickListener: BookmarkAdapter.OnItemClickListener<Bookmark>
+    ) {
+        itemView.txtBookmarkName.text = bookmark.name
```
This only works if the layout used contains the `TextView` `txtBookmarkName`. You should be binding the `bookmark.name` property to the `TextView`

puneet-pdx replied:
if I bind the `bookmark.name` to the `TextView` it is less cleaner than this. I have made the change so you can have a look at how I am doing it. 
If the User want to override the layout then they can name the `TextView` in the layout as `txtBookmarkName` to make use of this binding, right ?

LDuncAndroid replied:
But this isn't databinding in the same sense. The Adapter should allow the user to bind whatever properties from the Bookmark they wish to bind to their view. It just so happens that we're only binding the name here. By using a pure Data bound layout, we're showing the user how they can also do so.

LDuncAndroid commented on arcgis-android-toolkit/src/main/java/com/esri/arcgisruntime/toolkit/bookmark/BookmarkView.kt:
```diff
@@ -0,0 +1,75 @@
+
+class BookmarkView : FrameLayout {
+
+    var recyclerView: RecyclerView? = null
```
As you are inflating the layout in `init()`, this property can be a `val` that is lazily instantiated.

puneet-pdx replied:
updated

LDuncAndroid commented on arcgis-android-toolkit/src/main/java/com/esri/arcgisruntime/toolkit/bookmark/BookmarkView.kt:
```diff
@@ -0,0 +1,75 @@
+
+    var recyclerView: RecyclerView? = null
+
+    var bookmarksAdapter: BookmarkAdapter? = null
```
This can be instantiated here.

puneet-pdx replied:
done

LDuncAndroid commented on arcgis-android-toolkit/src/main/java/com/esri/arcgisruntime/toolkit/bookmark/BookmarkView.kt:
```diff
@@ -0,0 +1,75 @@
+
+    var onItemClickListener: OnItemClickListener<Bookmark>? = null
+
+    interface OnItemClickListener<T> {
```
Is there any need for another interface?  if so, does it need to have a generic type `T` if we know the type of item is a `Bookmark`?

puneet-pdx replied:
we want the user to set the listener on the `BookmarkView` .  I have updated the type to `Bookmark`

LDuncAndroid replied:
I was suggesting re-using `BookmarkAdapter.OnItemClickListener` instead but leave it here if you think it's cleaner.

LDuncAndroid commented on arcgis-android-toolkit/src/main/java/com/esri/arcgisruntime/toolkit/bookmark/BookmarkView.kt:
```diff
@@ -0,0 +1,75 @@
+        if (recyclerView == null) {
+            recyclerView = _bookmarkRecyclerView
+        }
+        recyclerView?.layoutManager = LinearLayoutManager(context)
```
You can set the LayoutManager in XML: https://stackoverflow.com/a/35680716

puneet-pdx replied:
done

LDuncAndroid commented on arcgis-android-toolkit/src/main/java/com/esri/arcgisruntime/toolkit/bookmark/BookmarkView.kt:
```diff
@@ -0,0 +1,75 @@
+
+    var bookmarksAdapter: BookmarkAdapter? = null
+
+    var onItemClickListener: OnItemClickListener<Bookmark>? = null
```
By instantiating this here you can remove the nullability.

puneet-pdx replied:
this listener will be set by the user... how will I be instantiating it here ? can you explain?

LDuncAndroid replied:
Yeah, apologies, just realised what it was for.

LDuncAndroid commented on arcgis-android-toolkit/src/main/java/com/esri/arcgisruntime/toolkit/bookmark/BookmarkView.kt:
```diff
@@ -0,0 +1,75 @@
+
+    private fun init(context: Context) {
+        inflate(context, R.layout.layout_bookmarkview, this)
+        if (recyclerView == null) {
```
By lazily instantiating the property you can remove the nullability and null check.

puneet-pdx replied:
updated

LDuncAndroid commented on arcgis-android-toolkit/src/main/res/layout/item_bookmark_row.xml:
```diff
@@ -0,0 +1,26 @@
+  ~ limitations under the License.
+  -->
+
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Should be a bindable layout.

puneet-pdx replied:
can you explain why ? the binding is working in the `BookmarkAdapter.ViewHolder.bind()` method without this being a bindable layout..

LDuncAndroid replied:
But this isn't true data binding, explained here: https://github.com/Esri/arcgis-runtime-toolkit-android/pull/146/files/e451f2f85f26b6f0a9db0d8122a647217b3911e1#diff-3852308ac2ab80c12bbc863fe4507601R72

LDuncAndroid commented on arcgis-android-toolkit/src/main/res/layout/layout_bookmarkview.xml:
```diff
@@ -0,0 +1,27 @@
+  ~ limitations under the License.
+  -->
+
+<FrameLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Do we need the `FrameLayout` parent?

puneet-pdx replied:
not really, it works without it. But we have done it for other toolkit components, so we will do it for Bookmarks as well.

LDuncAndroid replied:
It's only used in `ArcGISArView` because we need the two views stacked on each other. If it can be removed it's more performant to do so.

LDuncAndroid commented on toolkit-test-app/src/main/java/com/esri/arcgisruntime/toolkit/test/bookmark/map/MapViewModel.kt:
```diff
@@ -0,0 +1,44 @@
+class MapViewModel(map: ArcGISMap) : ViewModel() {
+
+    private val _mapData: MutableLiveData<ArcGISMap> = MutableLiveData(map)
+    val mapData: LiveData<ArcGISMap> = _mapData
```
This isn't being used. I think you should be observing this and setting it on the `MapView` opposed to what you are doing here:

https://github.com/Esri/arcgis-runtime-toolkit-android/pull/146/files#diff-85eb21d63939d7e4be2fbe13c15ec13eR47

puneet-pdx replied:
updated

LDuncAndroid commented on toolkit-test-app/src/main/java/com/esri/arcgisruntime/toolkit/test/bookmark/BookmarkActivity.kt:
```diff
@@ -0,0 +1,59 @@
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_bookmark)
+
+        val portal = Portal("https://arcgisruntime.maps.arcgis.com/")
```
Can you perform this instantiation before `onCreate()`? Perhaps as another lazy property passing it to the `MapViewModel.Factory`?

puneet-pdx replied:
yes. Updated

puneet-pdx commented:
@LDuncAndroid I have addressed your comments. Please have another look.

LDuncAndroid commented on arcgis-android-toolkit/src/main/res/layout/layout_bookmarkview.xml:
```diff
@@ -0,0 +1,27 @@
+        android:layout_height="match_parent">
+
+    <androidx.recyclerview.widget.RecyclerView
+            android:id="@+id/_bookmarkRecyclerView"
```
Does this need an underscore? Unless `bookmarkRecyclerView` is used elsewhere in the View, it shouldn't be necessary.

puneet-pdx replied:
updated

LDuncAndroid commented on arcgis-android-toolkit/src/main/java/com/esri/arcgisruntime/toolkit/bookmark/BookmarkAdapter.kt:
```diff
@@ -0,0 +1,78 @@
+        bookmark: Bookmark,
+        onItemClickListener: BookmarkAdapter.OnItemClickListener<Bookmark>
+    ) {
+        if (itemView is ViewGroup && itemView.getChildAt(0) is TextView) {
```
This now forces the user to only use a `ViewGroup` if they'd used just a `TextView` this will not show the bookmark name. The solution I had in the adapter bound the Bookmark item to the `ViewDataBinding` instance and I felt it was cleaner than this.

LDuncAndroid commented on arcgis-android-toolkit/src/main/res/layout/layout_bookmarkview.xml:
```diff
@@ -0,0 +1,27 @@
+        android:layout_width="match_parent"
+        android:layout_height="match_parent">
+
+    <androidx.recyclerview.widget.RecyclerView
```
You still need to set the LayoutManager here: https://github.com/Esri/arcgis-runtime-toolkit-android/pull/146#discussion_r342482968

puneet-pdx replied:
ah yes, forgot to push this change. Updated

LDuncAndroid commented on arcgis-android-toolkit/build.gradle:
```diff
@@ -55,6 +56,9 @@ android {
     kotlinOptions {
         jvmTarget = '1.8'
     }
+    dataBinding {
```
This actually works without databinding enabled at the moment.

puneet-pdx replied:
No. We need dataBinding enabled to be able to use DatabindingUtil and ViewDataBinding classes in BookmarkAdapter.

https://github.com/Esri/arcgis-runtime-toolkit-android/pull/146/files#diff-6972c309ea40b7e84309ccee57d3091cR106

https://github.com/Esri/arcgis-runtime-toolkit-android/pull/146/files#diff-6972c309ea40b7e84309ccee57d3091cR135

LDuncAndroid commented:
Assigning back to you @puneet-pdx

LDuncAndroid commented:
Here's a cleaner example of mine, without using Data Binding to bind the OnClickListener: https://devtopia.esri.com/luk10114/Toolkit-Components/commit/4873294092a8b68fc04d9d6b2f012ffbd1942a3b

But notice that we are still providing the binding ID for the bookmark item, to allow the user to create their own layout without the constraints shown here.

puneet-pdx commented:
@LDuncAndroid I have addressed your comments.

@gunt0001 I have addressed your comments as well,
>I have looked over your PR 146. Overall looks good, in addition to Luke's comments I would add:
>- BookmarkAdapter could be a private nested class in BookmarkView for now so it's clear to users that they cannot instantiate one themselves
>- since we decided not to support layout customization of bookmark items for now, I don't think you need to pass in the layout res to the BookmarkAdapter ctor
>- The layout/styling of BookmarkActivity could do with some refinement, but that isn't a priority, can be done once the bookmark component is dialed in

Please have another look.

LDuncAndroid commented on arcgis-android-toolkit/src/main/java/com/esri/arcgisruntime/toolkit/bookmark/BookmarkView.kt:
```diff
@@ -0,0 +1,127 @@
+
+class BookmarkView : FrameLayout {
+
+    val recyclerView: RecyclerView by lazy { bookmarkRecyclerView }
```
Can this be private?

LDuncAndroid commented on arcgis-android-toolkit/src/main/java/com/esri/arcgisruntime/toolkit/bookmark/BookmarkView.kt:
```diff
@@ -0,0 +1,127 @@
+
+    private class BookmarkAdapter(
+        private val onItemClickListener: OnItemClickListener<Bookmark>,
+        diffCallback: DiffUtil.ItemCallback<Bookmark> = DiffCallback()
```
Could remove this from the `BookmarkAdapter` constructor since we're not allowing users to instantiate the adapter themselves.

```
    private class BookmarkAdapter(
        private val onItemClickListener: OnItemClickListener<Bookmark>
    ) : ListAdapter<Bookmark, ViewHolder>(DiffCallback())
```

LDuncAndroid commented on arcgis-android-toolkit/src/main/java/com/esri/arcgisruntime/toolkit/bookmark/BookmarkView.kt:
```diff
@@ -0,0 +1,127 @@
+        override fun onBindViewHolder(holder: ViewHolder, position: Int) =
+            holder.bind(getItem(position), onItemClickListener)
+
+        class DiffCallback : DiffUtil.ItemCallback<Bookmark>() {
```
Could be private.

LDuncAndroid commented:
Couple of things I spotted after approval.

puneet-pdx commented:
@gunt0001 can you please review. For testing I have added the test activity that shows bookmarks that are present in the web map being loaded.

puneet-pdx commented:
@alan-edi can you please review?

changanxian commented on toolkit-test-app/src/main/java/com/esri/arcgisruntime/toolkit/test/bookmark/BookmarksRepository.kt:
```diff
@@ -0,0 +1,45 @@
+import androidx.lifecycle.MutableLiveData
+import com.esri.arcgisruntime.loadable.LoadStatus
+import com.esri.arcgisruntime.mapping.ArcGISMap
+import com.esri.arcgisruntime.mapping.Bookmark
```
the `import`  is not used in the class, it should be removed

puneet-pdx commented:
@changanxian thanks. @alan-edi can you do the second review.

gunt0001 commented on arcgis-android-toolkit/src/main/java/com/esri/arcgisruntime/toolkit/bookmark/BookmarkView.kt:
```diff
@@ -0,0 +1,147 @@
+            override fun onItemClick(item: Bookmark) {
+                onItemClickListener?.onItemClick(item)
+            }
+        })
```
Is there a need to pass an `OnItemClickListener` to the ctor of BookmarkAdapter? It might be easier to simply instantiate this in the BookmarkAdapter ctor and then invoke the `onItemClickListener` on the outer class from there, if it was set.

puneet-pdx replied:
yes there is a need to pass it, as the `BookmarkAdapter` is a nested class and not an inner class so it does not know about `onItemClickListener` in the outer class. 
Cannot make `BookmarkAdapter` an inner class and initialize `onItemClickListener` in the inner class. 
```
object onItemClickListener2: OnItemClickListener<Bookmark> {
            override fun onItemClick(item: Bookmark) {
                onItemClickListener?.onItemClick(item)
            }
        }
```  
https://kotlinlang.org/docs/reference/object-declarations.html
`NOTE: object declarations can't be local (i.e. be nested directly inside a function), but they can be nested into other object declarations or non-inner classes.`

gunt0001 replied:
I think you could make `BookmarkAdapter` an inner class which could directly reference the `onItemClickListener` of BookmarkView. It would look something like this:
```
private inner class BookmarkAdapter() : ListAdapter<Bookmark, ViewHolder>(DiffCallback()) {

        private val onItemClickListener = object: OnItemClickListener<Bookmark> {
            override fun onItemClick(item: Bookmark) {
                this@BookmarkView.onItemClickListener?.onItemClick(item)
            }
        }
```
I would think this would make the code slightly simpler as BookmarkAdapter becomes a bit more encapsulated, i.e. you don't need to pass a click listener to its constructor.

gunt0001 commented on arcgis-android-toolkit/src/main/res/layout/item_bookmark_row.xml:
```diff
@@ -0,0 +1,41 @@
+    </LinearLayout>
+
+
+</layout>
```
missing newline

puneet-pdx replied:
added

gunt0001 commented on arcgis-android-toolkit/src/main/java/com/esri/arcgisruntime/toolkit/bookmark/BookmarkView.kt:
```diff
@@ -0,0 +1,147 @@
+     *
+     * @since 100.7.0
+     */
+    fun submitBookmarkList(bookmarkList: BookmarkList) {
```
I think this could be a property instead:
```
var bookmarks?: BookmarkList
```
Then you should be able to directly bind this property in the layout to a property on a ViewModel that exposes a BookmarkList

puneet-pdx replied:
updated

gunt0001 commented on toolkit-test-app/src/main/java/com/esri/arcgisruntime/toolkit/test/bookmark/BookmarkActivity.kt:
```diff
@@ -0,0 +1,67 @@
+            it.let {
+                mapView.map = mapViewModel.mapData.value
+            }
+        })
```
Could set the map on the MapView via data binding expression in the layout?

gunt0001 commented on toolkit-test-app/src/main/java/com/esri/arcgisruntime/toolkit/test/bookmark/BookmarkActivity.kt:
```diff
@@ -0,0 +1,67 @@
+
+        mapViewModel.bookmarks.observe(this, Observer {
+            it?.let { bookmarkView.submitBookmarkList(it) }
+        })
```
Could set the bookmarks on the BookmarkView via data binding expression in the layout?

puneet-pdx replied:
updated.

gunt0001 commented on toolkit-test-app/src/main/java/com/esri/arcgisruntime/toolkit/test/bookmark/BookmarksRepository.kt:
```diff
@@ -0,0 +1,44 @@
+import com.esri.arcgisruntime.mapping.ArcGISMap
+import com.esri.arcgisruntime.mapping.BookmarkList
+
+class BookmarksRepository(private val map: ArcGISMap) {
```
Should we ommit the repository here for simplicity? I don't think it adds much for this simple test/demo.

puneet-pdx replied:
removed

gunt0001 commented on toolkit-test-app/src/main/res/layout/activity_bookmark.xml:
```diff
@@ -0,0 +1,48 @@
+    android:layout_width="match_parent"
+    android:layout_height="match_parent">
+
+    <com.esri.arcgisruntime.mapping.view.MapView
```
Would be good to do the data binding to the map here

puneet-pdx replied:
done

gunt0001 commented on toolkit-test-app/src/main/res/layout/activity_bookmark.xml:
```diff
@@ -0,0 +1,48 @@
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toTopOf="parent" />
+
+    <com.esri.arcgisruntime.toolkit.bookmark.BookmarkView
```
Would be good to do the data binding to the bookmarks here

puneet-pdx replied:
done

puneet-pdx commented on arcgis-android-toolkit/src/main/java/com/esri/arcgisruntime/toolkit/bookmark/BookmarkView.kt:
```diff
@@ -0,0 +1,147 @@
+
+    val recyclerView: RecyclerView by lazy { bookmarkRecyclerView }
+
+    private val bookmarksAdapter: BookmarkAdapter = BookmarkAdapter(
```
It is not getting initialized by a lambda, so I don't think `by lazy` applies. 
Cannot do a `lateinit` as it is a `val` and not a `var`

puneet-pdx commented:
@gunt0001 I have addressed your comments. Please have another look.

gunt0001 commented on arcgis-android-toolkit/src/main/java/com/esri/arcgisruntime/toolkit/bookmark/BookmarkView.kt:
```diff
@@ -0,0 +1,144 @@
+        value?.let {
+            field = value
+            bookmarksAdapter.submitList(value)
+        }
```
if `null` is passed for `value` I think you should clear any bookmarks set on `bookmarksAdapter`. In other words, I don't think you need to check for non null using `let`

puneet-pdx replied:
updated

gunt0001 commented on toolkit-test-app/src/main/java/com/esri/arcgisruntime/toolkit/test/bookmark/map/MapViewModel.kt:
```diff
@@ -0,0 +1,55 @@
+    val map: ArcGISMap by lazy {
+        val portal = Portal("https://arcgisruntime.maps.arcgis.com/")
+        val portalItem = PortalItem(portal, "e1aa3973d50a456f998406a7c4dfd804")
+        portal.credential = UserCredential("ArcGISRuntimeSDK", "agsRT3dk")
```
Can we share this credential or better use a public map?

puneet-pdx replied:
updated

gunt0001 commented on toolkit-test-app/src/main/java/com/esri/arcgisruntime/toolkit/test/bookmark/map/MapViewModel.kt:
```diff
@@ -0,0 +1,55 @@
+class MapViewModel : ViewModel() {
+
+    val map: ArcGISMap by lazy {
+        val portal = Portal("https://arcgisruntime.maps.arcgis.com/")
```
Is this a portal for public use?

puneet-pdx replied:
yes

gunt0001 commented on toolkit-test-app/src/main/java/com/esri/arcgisruntime/toolkit/test/bookmark/map/MapViewModel.kt:
```diff
@@ -0,0 +1,55 @@
+    private val _bookmarks: MutableLiveData<BookmarkList> = MutableLiveData()
+    val bookmarks: LiveData<BookmarkList>
+        get() {
+            loadBookmarks()
```
I don't think there is a need for this call here. On creation of the MapViewModel you can listen to the loading of the map and once it has loaded you call `_bookmarks.postValue(map.bookmarks)`

gunt0001 commented on arcgis-android-toolkit/src/main/java/com/esri/arcgisruntime/toolkit/bookmark/BookmarkView.kt:
```diff
@@ -0,0 +1,144 @@
+            }
+        })
+
+    var bookmarkList: BookmarkList? = null
```
I suggest we call the property `bookmarks` instead of `bookmarkList`

puneet-pdx replied:
updated

puneet-pdx commented:
@gunt0001 I have addressed the comments.

gunt0001 commented on toolkit-test-app/src/main/java/com/esri/arcgisruntime/toolkit/test/bookmark/BookmarkActivity.kt:
```diff
@@ -0,0 +1,52 @@
+
+        mapViewModel.map.addDoneLoadingListener{
+            mapViewModel.updateBookmarks()
+        }
```
I guess the MapViewModel itself could listen to the map's load status and then update the bookmarks. I'll leave it to you whether you want to do this here or in MapViewModel.

puneet-pdx replied:
I have moved the addDoneLoadingListener to MapViewModel
